include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -1497,6 +1498,213 @@ MatrixFree<dim, Number, VectorizedArrayType>::initialize_indices(
+    for (const auto &di : dof_info)
+      for (const auto &v : di.index_storage_variants[2])
+        is_non_buffering_sm_supported &=
+          (v == internal::MatrixFreeFunctions::DoFInfo::IndexStorageVariants::
+                  contiguous);
+
+    // TODO: get from TaskInfo
+    const MPI_Comm communicator_sm = MPI_COMM_SELF;
+
+    if (is_non_buffering_sm_supported)
+      {
+        // TODO: move into Utilities::MPI
+        const auto procs_of_sm = [](const MPI_Comm &comm,
+                                    const MPI_Comm &comm_shared) {
+          const unsigned int rank = Utilities::MPI::this_mpi_process(comm);
+          const unsigned int size =
+            Utilities::MPI::n_mpi_processes(comm_shared);
+
+          std::vector<unsigned int> ranks(size);
+          MPI_Allgather(
+            &rank, 1, MPI_UNSIGNED, ranks.data(), 1, MPI_INT, comm_shared);
+
+          return ranks;
+        };
+
+        // gather the ranks of the shared-memory domain
+        const std::vector<unsigned int> sm_procs =
+          procs_of_sm(task_info.communicator, communicator_sm);
+
+        // get my rank within the shared-memory domain
+        const unsigned int my_sm_pid = std::distance(
+          sm_procs.begin(),
+          std::find(sm_procs.begin(), sm_procs.end(), task_info.my_pid));
+
+        // process any dof_info
+        const auto  di_index = 0;
+        const auto &di       = dof_info[di_index];
+
+        std::array<std::vector<std::pair<unsigned int, unsigned int>>, 3>
+          cell_indices_contiguous_sm;
+
+        // collect for each cell: sm rank + offset
+        cell_indices_contiguous_sm[2] = [&]() {
+          // result
+          std::vector<std::pair<unsigned int, unsigned int>> cells(
+            di.dof_indices_contiguous[2].size(),
+            {numbers::invalid_unsigned_int, numbers::invalid_unsigned_int});
+
+          // locally-owned cells
+          std::vector<std::pair<unsigned int, types::global_cell_index>>
+            cells_locally_owned;
+          cells_locally_owned.reserve(n_cell_batches() * n_lanes);
+
+          // shared-ghost cells (categorized according to sm-rank)
+          std::vector<
+            std::vector<std::pair<unsigned int, types::global_cell_index>>>
+            cells_shared_ghosts(sm_procs.size());
+
+          for (unsigned int cell = 0;
+               cell < n_cell_batches() + n_ghost_cell_batches();
+               ++cell)
+            for (unsigned int v = 0; v < this->n_components_filled(cell); ++v)
+              {
+                const unsigned int index = cell * n_lanes + v;
+
+                // determine global cell index
+                const unsigned int local_dof_index =
+                  di.dof_indices_contiguous[2][index];
+                const types::global_cell_index global_cell_index =
+                  di.vector_partitioner->local_to_global(local_dof_index) /
+                  this->get_dofs_per_cell(di_index);
```

Comment:
I guess it would be more safe to use `CellAccessor::global_active_cell_index()` and `CellAccessor::global_level_cell_index()` from #10490.